Allow WebAuthn .create() to wait for previous cancel requests to finish

abigailfox opened the pull request:
#### ded025f42cdea7d5f25c7125ce03c5a95f0a696f
<pre>
Allow WebAuthn .create() to wait for previous cancel requests to finish
<a href="https://bugs.webkit.org/show_bug.cgi?id=257176">https://bugs.webkit.org/show_bug.cgi?id=257176</a>
<a href="https://rdar.apple.com/109936742">rdar://109936742</a>

Reviewed by Pascoe.

Allow WebAuthn .create() to monitor and wait for previous cancel requests to finish, to avoid conflicts

* Source/WebCore/Modules/webauthn/AuthenticatorCoordinator.cpp:
(WebCore::AuthenticatorCoordinator::create):
(WebCore::AuthenticatorCoordinator::discoverFromExternalSource):
* Source/WebCore/Modules/webauthn/AuthenticatorCoordinator.h:
* Source/WebCore/Modules/webauthn/AuthenticatorCoordinatorClient.h:
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
(WebKit::WebAuthenticatorCoordinatorProxy::performRequest):
* Source/WebKit/UIProcess/WebAuthentication/WebAuthenticatorCoordinatorProxy.cpp:
(WebKit::WebAuthenticatorCoordinatorProxy::~WebAuthenticatorCoordinatorProxy):
(WebKit::WebAuthenticatorCoordinatorProxy::cancel):
* Source/WebKit/UIProcess/WebAuthentication/WebAuthenticatorCoordinatorProxy.h:
* Source/WebKit/UIProcess/WebAuthentication/WebAuthenticatorCoordinatorProxy.messages.in:
* Source/WebKit/WebProcess/WebAuthentication/WebAuthenticatorCoordinator.cpp:
(WebKit::WebAuthenticatorCoordinator::cancel):
* Source/WebKit/WebProcess/WebAuthentication/WebAuthenticatorCoordinator.h:

Canonical link: <a href="https://commits.webkit.org/273918@main">https://commits.webkit.org/273918@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0aa58c470b64dc92f2eab2b407fa3c40d536ad94

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/36932 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/15857 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/39197 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/39550 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/33048 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/38220 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/18383 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/12980 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/31580 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/37494 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/13386 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/32604 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/11677 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/11700 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/32900 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/40802 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/33491 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/33248 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/37601 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/11989 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/9798 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/35728 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/13642 "Built successfully") | | | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/8400 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/12370 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/12898 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jyavenard commented on Source/WebKit/WebProcess/WebAuthentication/WebAuthenticatorCoordinator.cpp:
```diff
@@ -95,9 +95,9 @@ void WebAuthenticatorCoordinator::isUserVerifyingPlatformAuthenticatorAvailable(
+void WebAuthenticatorCoordinator::cancel(CompletionHandler<void()>&& handler)
 {
-    m_webPage.send(Messages::WebAuthenticatorCoordinatorProxy::Cancel());
+    m_webPage.sendWithAsyncReply(Messages::WebAuthenticatorCoordinatorProxy::Cancel(), handler);
```
WTFMove(handler)

pascoej commented on Source/WebCore/Modules/webauthn/AuthenticatorCoordinator.cpp:
```diff
@@ -191,6 +191,17 @@ void AuthenticatorCoordinator::create(const Document& document, CredentialCreati
     };
+
+    if (m_isCancelling) {
+        m_queuedRequest = [weakThis = WeakPtr { *this }, weakFrame = WeakPtr { *frame }, weakDocument = WeakPtr<Document, WeakPtrImplWithEventTargetData> { document }, clientDataJsonHash = WTFMove(clientDataJsonHash), createOptions = WTFMove(createOptions), callback = WTFMove(callback)]() mutable {
```
I think we should take in `document->securityOrigin() `(`callerOrigin`) here instead of getting in inside the lambda.

pascoej commented on Source/WebCore/Modules/webauthn/AuthenticatorCoordinator.cpp:
```diff
@@ -251,10 +262,18 @@ void AuthenticatorCoordinator::discoverFromExternalSource(const Document& docume
 
     if (requestOptions.signal) {
-        requestOptions.signal->addAlgorithm([weakThis = WeakPtr { *this }](JSC::JSValue) {
+        m_isCancelling = true;
```
This isn't the point where we are cancelling, only that the RP passed an abort signal. This needs to go in the lambda passed to addAlgorithm below.

pascoej replied:
ditto

pascoej commented on Source/WebCore/Modules/webauthn/AuthenticatorCoordinator.cpp:
```diff
@@ -251,10 +262,18 @@ void AuthenticatorCoordinator::discoverFromExternalSource(const Document& docume
+                if (!weakThis)
+                    return;
+                if (auto queuedRequest = WTFMove(weakThis->m_queuedRequest)) {
+                    weakThis->m_isCancelling = false;
```
This and the line below should be indented at same level (maybe they are and the GitHub UI is confusing.)

abigailfox replied:
Indented at the same level as the lambda initiation? Or both ifs equal to each other (the latter is already true).

pascoej replied:
ditto

pascoej replied:
ditto

pascoej commented on Source/WebCore/Modules/webauthn/AuthenticatorCoordinator.cpp:
```diff
@@ -272,6 +291,16 @@ void AuthenticatorCoordinator::discoverFromExternalSource(const Document& docume
+            if (!weakThis || !weakFrame || !weakDocument)
+                return;
+            const auto& options = requestOptions.publicKey.value();
+            weakThis->m_client->getAssertion(*weakFrame, weakDocument->securityOrigin(), clientDataJsonHash, options, requestOptions.mediation, scopeAndCrossOriginParent, WTFMove(callback));
```
similar to above, let's capture the securityOrigin directly instead of capturing document and getting it in the lambda.

pascoej commented on Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
```diff
@@ -326,6 +326,10 @@ - (void)authorizationController:(ASAuthorizationController *)controller didCompl
                 response.signature = toArrayBuffer(credential.get().signature);
                 response.userHandle = toArrayBuffer(credential.get().userID);
             }
+            if (m_cancelHandler) {
```
I think you can do the fancy `if (auto cancelHandler = WTFMove(m_cancelHandler))` here.

pascoej commented on Source/WebKit/UIProcess/WebAuthentication/WebAuthenticatorCoordinatorProxy.cpp:
```diff
@@ -127,7 +127,7 @@ void WebAuthenticatorCoordinatorProxy::handleRequest(WebAuthenticationRequestDat
 
 #if !HAVE(UNIFIED_ASC_AUTH_UI)
-void WebAuthenticatorCoordinatorProxy::cancel()
+void WebAuthenticatorCoordinatorProxy::cancel(CompletionHandler<void()>&&)
```
Call the completion handler here (it might be important for test)

pascoej commented on Source/WebCore/Modules/webauthn/AuthenticatorCoordinator.cpp:
```diff
@@ -191,6 +191,17 @@ void AuthenticatorCoordinator::create(const Document& document, CredentialCreati
         ASSERT(!exception.message.isNull());
         promise.reject(exception.toException());
     };
+
```
Let's handle a block to handle creates being cancelled as well similar as the `if (requestOptions.signal) {` block below. The signal is`createOptions.signal.`

pascoej reviewed (changes requested):
Above changes

cdumez commented on Source/WebCore/Modules/webauthn/AuthenticatorCoordinator.cpp:
```diff
@@ -191,6 +207,17 @@ void AuthenticatorCoordinator::create(const Document& document, CredentialCreati
+                return;
+            const auto& options = createOptions.publicKey.value();
+            const auto& callerOrigin = weakFrame->document()->securityOrigin();
+            weakThis->m_client->makeCredential(*weakFrame, callerOrigin, clientDataJsonHash, options, createOptions.mediation, WTFMove(callback));
```
Please store the Frame in a RefPtr before passing it as parameter here, as per our smart pointer guidelines:
```
RefPtr frame = weakFrame.get();
if (!frame)
    return;
weakThis->m_client->makeCredential(*frame, callerOrigin, clientDataJsonHash, options, createOptions.mediation, WTFMove(callback));
```

cdumez commented on Source/WebCore/Modules/webauthn/AuthenticatorCoordinator.cpp:
```diff
@@ -272,6 +307,16 @@ void AuthenticatorCoordinator::discoverFromExternalSource(const Document& docume
+            if (!weakThis || !weakFrame)
+                return;
+            const auto& options = requestOptions.publicKey.value();
+            weakThis->m_client->getAssertion(*weakFrame, weakFrame->document()->securityOrigin(), clientDataJsonHash, options, requestOptions.mediation, scopeAndCrossOriginParent, WTFMove(callback));
```
Same comment about weakFrame.

cdumez replied:
Also `document()->protectedSecurityOrigin()` to make sure we protect the security origin we pass.

cdumez commented on Source/WebCore/Modules/webauthn/AuthenticatorCoordinator.cpp:
```diff
@@ -191,6 +207,17 @@ void AuthenticatorCoordinator::create(const Document& document, CredentialCreati
+            if (!weakThis || !weakFrame)
+                return;
+            const auto& options = createOptions.publicKey.value();
+            const auto& callerOrigin = weakFrame->document()->securityOrigin();
```
Please store in a `Ref`. We need a strong pointer on the stack, per our smart pointer guidelines since you're passing it as argument to another function.

pascoej commented on Source/WebCore/Modules/webauthn/AuthenticatorCoordinator.cpp:
```diff
@@ -177,6 +177,22 @@ void AuthenticatorCoordinator::create(const Document& document, CredentialCreati
     }
 
+    if (createOptions.signal) {
+        m_isCancelling = true;
```
This will set isCancelling to true whenever the site passes a signal. We want to set isCancelling to true whenever they call abort the signal, then cancel the request and continue it after the cancel has completed.

i.e. This `m_isCancelling = true;` call should go in the lambda passed to addAlgorithm, such that it gets called whenever the signal handler does, after the weakThis check on lines 183-184.

pascoej reviewed (changes requested):
This is close. The signal handlers still need fixed up.

pascoej commented on Source/WebCore/Modules/webauthn/AuthenticatorCoordinator.cpp:
```diff
@@ -177,6 +177,22 @@ void AuthenticatorCoordinator::create(const Document& document, CredentialCreati
 
+    if (createOptions.signal) {
+        createOptions.signal->addAlgorithm([weakThis = WeakPtr { *this }](JSC::JSValue) mutable {
+            m_isCancelling = true;
```
I don't think this is going to build. Put this below the check and access it with weakThis->m_isCancelling = true.

pascoej commented on Source/WebCore/Modules/webauthn/AuthenticatorCoordinator.cpp:
```diff
@@ -177,6 +177,22 @@ void AuthenticatorCoordinator::create(const Document& document, CredentialCreati
+            weakThis->m_client->cancel([weakThis = WTFMove(weakThis)] () mutable {
+                if (!weakThis)
+                    return;
+                weakThis->m_isCancelling = true;
```
This only gets called after the cancel is complete, this needs to be inside the signal block (happens whenever the rp calls the abortSignal) but outside the cancel completion handler (happens when the cancel is complete.)

abigailfox committed:
Allow WebAuthn .create() to wait for previous cancel requests to finish
https://bugs.webkit.org/show_bug.cgi?id=257176
rdar://109936742

Reviewed by Pascoe.

Allow WebAuthn .create() to monitor and wait for previous cancel requests to finish, to avoid conflicts

* Source/WebCore/Modules/webauthn/AuthenticatorCoordinator.cpp:
(WebCore::AuthenticatorCoordinator::create):
(WebCore::AuthenticatorCoordinator::discoverFromExternalSource):
* Source/WebCore/Modules/webauthn/AuthenticatorCoordinator.h:
* Source/WebCore/Modules/webauthn/AuthenticatorCoordinatorClient.h:
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
(WebKit::WebAuthenticatorCoordinatorProxy::performRequest):
* Source/WebKit/UIProcess/WebAuthentication/WebAuthenticatorCoordinatorProxy.cpp:
(WebKit::WebAuthenticatorCoordinatorProxy::~WebAuthenticatorCoordinatorProxy):
(WebKit::WebAuthenticatorCoordinatorProxy::cancel):
* Source/WebKit/UIProcess/WebAuthentication/WebAuthenticatorCoordinatorProxy.h:
* Source/WebKit/UIProcess/WebAuthentication/WebAuthenticatorCoordinatorProxy.messages.in:
* Source/WebKit/WebProcess/WebAuthentication/WebAuthenticatorCoordinator.cpp:
(WebKit::WebAuthenticatorCoordinator::cancel):
* Source/WebKit/WebProcess/WebAuthentication/WebAuthenticatorCoordinator.h:

Canonical link: https://commits.webkit.org/273918@main
